vmm/src/memory_manager.rs

```diff
@@ -496,23 +498,42 @@ impl MemoryManager {
         Ok((mem_regions, memory_zones))
     }
 
-    fn fill_saved_regions(&mut self, saved_regions: Vec<MemoryRegion>) -> Result<(), Error> {
-        for region in saved_regions {
-            if let Some(content) = region.content {
-                // Open (read only) the snapshot file for the given region.
-                let mut memory_region_file = OpenOptions::new()
-                    .read(true)
-                    .open(content)
-                    .map_err(Error::SnapshotOpen)?;
-
-                self.guest_memory
-                    .memory()
-                    .read_exact_from(
-                        GuestAddress(region.start_addr),
-                        &mut memory_region_file,
-                        region.size as usize,
+    fn fill_saved_regions(
+        &mut self,
+        file_path: PathBuf,
+        saved_regions: MemoryRangeTable,
+    ) -> Result<(), Error> {
+        if saved_regions.is_empty() {
+            return Ok(());
+        }
+
+        // Open (read only) the snapshot file.
+        let mut memory_file = OpenOptions::new()
+            .read(true)
+            .open(file_path)
+            .map_err(Error::SnapshotOpen)?;
+
+        let guest_memory = self.guest_memory.memory();
+        for range in saved_regions.regions() {
+            let mut offset: u64 = 0;
+            // Here we are manually handling the retry in case we can't the
```

Comment:
Nit: "... can't `read` the whole region ..."

Reply:
Ah thanks for catching that!